### PR TITLE
CCC SW Mech

### PR DESCRIPTION
**Pull request by Adbritto**
The SW Mech code for CCC, Tested on individual subsystems, integrated testing in progress

### CODE REVIEW

**Comment by Adbritto** [comment]
Will add a commit soon to resolve Merge Conflicts

**Comment by Adbritto** [comment]
Last 2 commits to resolve merge conflict, no code change

File: src/main/deploy/pathplanner/paths/Copy of B3_C8.path
Comment (AG6GR) [inline src/main/deploy/pathplanner/paths/Copy of B3_C8.path:1]: Do we need this "Copy"?

File: src/main/java/frc/robot/Constants.java
Comment (AG6GR) [inline src/main/java/frc/robot/Constants.java]: For these constants I find it helpful to include the units in the names, i.e. `AUTO_REVING_SECS`, `FEED_SHOOTER_REV_PER_SEC`

File: src/main/java/frc/robot/HardwareMap.java
Comment (AG6GR) [inline src/main/java/frc/robot/HardwareMap.java]: These ID's don't match with the electrical team's Wiring Index document in the drive. I'd double check with them to either update the doc, or update the SPARK MAX configs.

File: src/main/java/frc/robot/TeleopInput.java
Comment (AG6GR) [inline src/main/java/frc/robot/TeleopInput.java]: I'm curious why we're using an analog joystick here? Personally I'd use the D pad

File: src/main/java/frc/robot/systems/ClimberMechFSM.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/ClimberMechFSM.java]: I feel like this would be much more readable as a series of simple if/else statements. Also consider making this something like`calculatePower(double currentPosition)`, since as far as I can tell you're always passing constants into `value` and `raisedPosition` anyway.

At a higher level, isn't this just a hardcoded version of the proportional (P) term of PID?

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java]: Can you put these DIO pin numbers into the HardwareMap?

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java:80]: Reminder that the onboard I2C port is known to cause RoboRIO freezes https://docs.wpilib.org/en/stable/docs/yearly-overview/known-issues.html#onboard-i2c-causing-system-lockups

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java]: Shouldn't you calculate this before putting in the dashboard? Also hasNote can be a function local variable, doesn't need to be a class variable.

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java]: I feel like a more appropriate names would along the lines of "approximatelyEquals"

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java:272]: Well technically this is just proportional control

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java:286]: WPILib also provides a PIDController class if you don't want to write up the math yourself for I and D. Although if P only is working for you, then no need to overcomplicate things. https://docs.wpilib.org/en/stable/docs/software/advanced-controls/controllers/pidcontroller.html

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java]: Is there a general description somewhere of what all the colors mean? Some quick reference we can hand to the drivers.

**Review by AG6GR — COMMENTED** [review]
Is there a way we can make use of the FSM rather than Commands in autonomous? I'd much rather each autonomous event trigger just set a specific FSM state. Then we can reuse the FSM handlers rather than coding an entirely new set of behaviors and handlers just for autonomous.

File: src/main/java/frc/robot/HardwareMap.java
Comment (Adbritto) [reply src/main/java/frc/robot/HardwareMap.java]: We can confirm and update these using the REV client on Monday

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (Adbritto) [reply src/main/java/frc/robot/systems/IntakeFSMSystem.java:286]: I believe we tried using the WPILib PIDController at one point, but we were having issues tuning it. We also don't really need to use the I or D so a P only controller works just fine. We could try using the CTRE MotionMagic controller after ccc as a project since we now have access to krakens.

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (Adbritto) [reply src/main/java/frc/robot/systems/IntakeFSMSystem.java]: We will be changing these based on driver preferences and then we will also document them with the robot control mapping

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (Adbritto) [reply src/main/java/frc/robot/systems/IntakeFSMSystem.java:80]: We are aware of this issue, but we weren't able to find time to test an alternative. We considered break beams but the mechanism is too shaky and could result in false positives.

File: src/main/deploy/pathplanner/paths/Copy of B3_C8.path
Comment (Adbritto) [reply src/main/deploy/pathplanner/paths/Copy of B3_C8.path:1]: @beranki is there a reason we made these copies?

File: src/main/java/frc/robot/TeleopInput.java
Comment (Adbritto) [reply src/main/java/frc/robot/TeleopInput.java]: Ideally we would change it according to driver preferences. That is of course when we do eventually decide upon who the drivers are

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (Adbritto) [reply src/main/java/frc/robot/systems/IntakeFSMSystem.java:272]: We figured that we wouldn't need to use the I and D so we just ended up going with proportional. I can update the javadoc but would you also want me to rename the method?

File: src/main/deploy/pathplanner/paths/Copy of B3_C8.path
Comment (Adbritto) [reply src/main/deploy/pathplanner/paths/Copy of B3_C8.path:1]: >Is there a way we can make use of the FSM rather than Commands in autonomous? I'd much rather each autonomous event trigger just set a specific FSM state. Then we can reuse the FSM handlers rather than coding an entirely new set of behaviors and handlers just for autonomous.

As for this, I believe we need to use Named Commands to trigger autonomous events with path planner but we can still make use of the FSM state handler methods if we wish to do so. But I'm not exactly sure if that's correct. 
@beranki  can talk more about path planner and autonomous as he wrote the initial code for it.

**Comment by Adbritto** [comment]
This commit contains all the changes recommended in the resolved comments

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (Adbritto) [reply src/main/java/frc/robot/systems/IntakeFSMSystem.java]: > Shouldn't you calculate this before putting in the dashboard?

Yes, I will update that with the next commit
>Also hasNote can be a function local variable, doesn't need to be a class variable.

The reason it is a class variable is so that we don't need to call the hasNote() function in every method that requires it and instead we just set it equal to the hasNote variable every loop of the update() function and call the variable instead for all other purposes. I think we started doing this after MBR because part of the sensor code had an overrun issue and we believe it was the hasNote() function being called repeatedly.

File: src/main/java/frc/robot/systems/ClimberMechFSM.java
Comment (Adbritto) [reply src/main/java/frc/robot/systems/ClimberMechFSM.java]: @queerboi and @Coldslicer would be able to comment on that, as they were primarily working on the climber

File: src/main/java/frc/robot/systems/ClimberMechFSM.java
Comment (thatSteveFan) [inline src/main/java/frc/robot/systems/ClimberMechFSM.java]: x and y are bad variable names here, be more specific on whatinput and output this is using.

File: src/main/java/frc/robot/systems/ClimberMechFSM.java
Comment (thatSteveFan) [reply src/main/java/frc/robot/systems/ClimberMechFSM.java]: @AG6GR I worked with Sharvil on this. The function here is supposed to give full power most of the time, then back off as we get towards the end (top or bottom depending on the direction). I told them to use a little step function just to make the code simpler.

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java]: There should only ever be one TeleopInput at Robot.java level. No need to create your own

File: src/main/java/frc/robot/systems/IntakeFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/IntakeFSMSystem.java]: You should be able to input.mechLeftRumble

File: src/main/java/frc/robot/systems/ShooterFSMSystem.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/ShooterFSMSystem.java]: Again, no need to create another TeleopInput

**Review by AG6GR — DISMISSED** [review]

**Comment by Adbritto** [comment]
I am holding off on the merge for now because we might make changes after the field test today. Will merge after

**Comment by Adbritto** [comment]
Changes to Auto Commands:

1. Add timers to any commands that rely on the color sensor detecting the note so we don't end up over intaking or so that we don't keep running the intake motors after we miss a note.
2. For the shooting commands both `indexerMotor` and `shooterMotor` will be run from within the `ShootNoteCommand` located in the `ShooterFSM`. This was done to avoid having to run multiple commands in parallel and so that we wont have to time each path differently in terms of when it starts and stops shooting. instead we can rev the shooter while moving and shoot the note out after the path ends.
3. The `pivotMotor` will be set to 0 in any commands except for when moving the deployer to ground or home.

**Review by AG6GR — APPROVED** [review]

**Pull request merged**
